runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/MapControlClientPool.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.fnexecution.control;
+
+import com.google.common.collect.Maps;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * A {@link ControlClientPool} backed by a client map. It is expected that a given client id will be
+ * requested at most once.
+ */
+public class MapControlClientPool implements ControlClientPool {
+
+  /** Creates a {@link MapControlClientPool} with an unspecified timeout. */
+  public static MapControlClientPool create() {
+    return withTimeout(Duration.ofSeconds(Long.MAX_VALUE));
+  }
+
+  /**
+   * Creates a {@link MapControlClientPool} with the given timeout. Timeouts only apply to source
+   * requests.
+   */
+  public static MapControlClientPool withTimeout(Duration timeout) {
+    return new MapControlClientPool(timeout);
+  }
+
+  private final Duration timeout;
+  private final Object lock = new Object();
+
+  @GuardedBy("lock")
+  private final Map<String, CompletableFuture<InstructionRequestHandler>> clients =
```

Comment:
<!--thread_id:cc_182896321_t; commit:1fd121da1417624b3b84f0300648251da64b9cb5; resolved:1-->
<!--section:context-quote-->
> **jkff** wrote:
> Why is the value type CompletableFuture, rather than simply InstructionRequestHandler? The get/put methods operate in terms of IRH, and the CompletableFuture we wrap it into seems trivial.

<!--section:body-->
Doing so without futures is not trivial due to the source and sink operating in different environments and due to the source call blocking until a client is available.

We could trivially implement a source that needs to be polled for an available client, but that would make for an unusable and inefficient interface.

If you're not convinced, I think we should talk offline. @tgroh and I spent some time whiteboarding earlier and concluded that this was the cleanest way to do what we want.